File: Grid Experiments/ShortestPath.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;
class Graph{
    public:
    void shortestPath(int V,vector<int> adj[],int dest)
    {
        // to print the list
        // for(int i=0;i<V;i++)
        // {
        //     for(auto it:adj[i])
        //     {
        //         cout<<it<<" ";
        //     }
        //     cout<<endl;
        // }
        vector<int> where(V);
        vector<int> cost(V);
        queue<int> q;
        vector<int> traversal;
        vector<int> visited(V,0);
        visited[0]=1;
        cost[0]=0;
        where[0]='#';
        q.push(0);
        int found=0;
        while(!q.empty())
        {
            if(found)break;
            int Y = q.front();
            q.pop();
            traversal.push_back(Y);
            for(auto it:adj[Y])
            {
                if(!visited[it])
                {
                    visited[it] =1;
                    q.push(it);
                    where[it] = Y;
                    cost[it] = cost[Y]+ 1;
                }
                if(it==dest)
                {
                    found=1;
                    break;
                }
            }
        }
        // for(auto i:where)
        // {
        //     cout<<i<<" ";
        // }
        // cout<<endl;
        // for(auto i:cost)
        // {
        //     cout<<i<<" ";
        // }
        // cout<<endl;
        // for(auto i:traversal)
        // {
        //     cout<<i<<" ";
        // }
        // cout<<endl;
        vector<int> path;
        int next=dest;
        while(where[next]!=35)
        {
            path.push_back(next);
            next = where[next];
        }
        path.push_back(0);
        reverse(path.begin(),path.end());
        for(auto i:path)
        {
            cout<<"->"<<i;
        }
        cout<<endl;

    }
        
};
int main()
{
    int V=16,E=22;
    vector<int> adj[V];
    for(int i=0;i<V;i++)
    {
        adj[i].push_back(i);
    }
    for(int i=0;i<E;i++)
    {
        int u,v;
        cin>>u>>v;
        adj[u].push_back(v);  
        adj[v].push_back(u);  

    }
    Graph g;

    g.shortestPath(V,adj,11);//passing destination only
    return 0;
}

File: Grid Experiments/shortestpath.js
var Q = require("./queue.js");
var userInput = [0 ,1 ,2 ,3 ,0 ,4 ,2 ,6 ,3 ,7 ,4 ,5, 5 ,6 ,6 ,7, 4, 8 ,5 ,9 ,6 ,10 ,7 ,11 ,8 ,9 ,9 ,10 ,10 ,11 ,8 ,12 ,9 ,13, 10 ,14, 11 ,15 ,12 ,13,13 ,14 ,14 ,15]
/*
q.push();
let ele = q.shift();
q.isEmpty();
q.front();
*/ 


function shortestPath(V,adj,src,dest)
{
    let where = new Array(V);
    let cost = new Array(V);
    var q = Q.createQueue(5);
    let traversal  = new Array(V);
    let visited = new Array(V).fill(0);
    visited[src]=1;
    cost[src]=0;
    where[src]=-1;
    q.push(src);
    let found=0;
    while(!q.isEmpty())
    {
        if(found)
        {
            break;
        }
        let Y = q.shift() // in this queue implementation, the popped element is returned as well
        traversal.push(Y);

        adj[Y].forEach(it => {
            if(!visited[it])
            {
                visited[it]=1;
                q.push(it);
                where[it] = Y;
                cost[it] = cost[Y] + 1;
            }
            if(it==dest)
            {
                found=1;
               
            }
        });
    }
    let next = dest;
    var path = [];
    while(where[next]!=-1)
    {
        path.push(next);
        next=where[next];

    }
    path.push(src);
    path.reverse();
    console.log(path);
}
//*MAIN 

let V= 16;
let E=22;
let adj = new Array;
for(let i=0;i<V;i++)
{
    let temp = new Array;
    temp.push(i);
    adj.push(temp);
}
for(let i=0;i<userInput.length;i+=2)
{
    let u = userInput[i];
    let v = userInput[i+1];
    adj[u].push(v);
    adj[v].push(u);
}
// console.log(adj);
shortestPath(V,adj,1,2);


